File: mfga-xposed/app/src/main/java/com/mfga/xposed/FontIdentity.java
package com.mfga.xposed;

/**
 * Single source of truth for the installed primary font's family name and path.
 *
 * <p>De-hardcoding: the family/path used to be repeated in every class that needed
 * them. Centralising here means swapping the installed font (e.g. a different
 * variable family) only touches one place. The values still match
 * {@code config/font-source.json} ({@code family} / {@code installedFile}); the
 * font packager remains the authority that verifies the file's hash and axes.
 */
public final class FontIdentity {
    /** Matches {@code family} in config/font-source.json. */
    public static final String FAMILY = "WenYuan Rounded SC VF";
    /** Matches {@code /system/fonts/} + {@code installedFile} in config/font-source.json. */
    public static final String FONT_PATH = "/system/fonts/Selffont-WenYuanRoundedSCVF.ttf";

    private FontIdentity() {}
}

File: mfga-xposed/app/src/main/kotlin/com/mfga/xposed/FontIdentity.kt
package com.mfga.xposed

/**
 * Single source of truth for the installed primary font family and visibility path.
 *
 * The public build uses the OFL-derived Selffont Maru family. Its five static
 * faces are selected by fonts.xml for Android weights 100..900; the regular face
 * is enough to prove that the family is visible in a Firefox process.
 */
object FontIdentity {
    /** Matches `family` in config/font-source.json. */
    const val FAMILY = "Selffont Maru"

    /** Matches the visibilityFile in config/font-source.json. */
    const val FONT_PATH = "/system/fonts/SelffontMaru-Regular.ttf"

    /** Static face paths emitted by fonts.xml; Android weights use nearest-face mapping. */
    private val FACE_PATHS = mapOf(
        300 to "/system/fonts/SelffontMaru-Light.ttf",
        400 to FONT_PATH,
        500 to "/system/fonts/SelffontMaru-Medium.ttf",
        700 to "/system/fonts/SelffontMaru-Bold.ttf",
        900 to "/system/fonts/SelffontMaru-Black.ttf",
    )

    fun facePathForWeight(weight: Int): String = FACE_PATHS.entries.minWithOrNull(
        compareBy<Map.Entry<Int, String>> { kotlin.math.abs(it.key - weight) }
            .thenByDescending { it.key }
    )!!.value

    /**
     * The verified no-visible-glyph metrics carrier whose nominal line metrics
     * each Selffont Maru face is normalised to. Matches `METRIC_CARRIER` in
     * tools/font_config.py.
     */
    const val CARRIER_PATH = "/system/fonts/Roboto-Regular.ttf"
}
